server/src/test/java/org/elasticsearch/indices/recovery/RecoverySettingsTests.java

```diff
@@ -100,6 +102,104 @@ public void testSnapshotDownloadPermitCanBeDynamicallyUpdated() {
         permit.close();
     }
 
+    public void testInsufficientNumberOfPermitsMessage() {
+        ClusterSettings clusterSettings = new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        RecoverySettings recoverySettings = new RecoverySettings(
+            Settings.builder()
+                .put(INDICES_RECOVERY_MAX_CONCURRENT_SNAPSHOT_FILE_DOWNLOADS_PER_NODE.getKey(), 5)
+                .put(CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_INCOMING_RECOVERIES_SETTING.getKey(), 2)
+                .put(INDICES_RECOVERY_MAX_CONCURRENT_SNAPSHOT_FILE_DOWNLOADS.getKey(), 3)
+                .build(),
+            clusterSettings
+        );
+
+        final String expectedMessage = String.format(
+            Locale.ROOT,
+            """
+                Unable to acquire permit to use snapshot files during recovery,
+                this recovery will recover index files from the source node.
+                Ensure snapshot files can be used during recovery by setting [%s] to be no greater than [2].
+                Current values of [%s] = [5], [%s] = [2]
+                """,
+            INDICES_RECOVERY_MAX_CONCURRENT_SNAPSHOT_FILE_DOWNLOADS.getKey(),
+            INDICES_RECOVERY_MAX_CONCURRENT_SNAPSHOT_FILE_DOWNLOADS_PER_NODE.getKey(),
+            CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_INCOMING_RECOVERIES_SETTING.getKey()
+        );
+
+        final Logger logger = LogManager.getLogger(RecoverySettings.class);
+        final MockLogAppender warnCaptureAppender = new MockLogAppender();
+        warnCaptureAppender.addExpectation(
+            new MockLogAppender.SeenEventExpectation("WARN-Capture", RecoverySettings.class.getCanonicalName(), Level.WARN, expectedMessage)
+        );
+
+        try {
+            warnCaptureAppender.start();
+            Loggers.addAppender(logger, warnCaptureAppender);
+
+            // Allow the first recovery to obtain a permit
+            Releasable permit = recoverySettings.tryAcquireSnapshotDownloadPermits();
+            assertThat(permit, is(notNullValue()));
+
+            // Deny the second recovery to get the permit
+            assertThat(recoverySettings.tryAcquireSnapshotDownloadPermits(), is(nullValue()));
+
+            // assert that WARN message was logged
+            warnCaptureAppender.assertAllExpectationsMatched();
+
+        } finally {
+            Loggers.removeAppender(logger, warnCaptureAppender);
+            warnCaptureAppender.stop();
+        }
```

Comment:
I have recently merged an utility that simplifies log capturing setup, you might consider using it here: https://github.com/elastic/elasticsearch/blob/3808b53914aeb1e48eb0941b1400bfc2c3d7d1c9/test/framework/src/main/java/org/elasticsearch/test/MockLogAppender.java#L268-L275